### ISSUE TITLE
Connectivitiy Preserve Modeler NodalSolutionStepVariablesList Check Method

### ISSUE THREAD

**Issue by sunethwarna** [KratosMultiphysics/Kratos#4979]
Hi,

After discussing with @jcotela about the #4897 and #4267 we went through the connectivity_preserve_modeler and came to following conclusions and suggestions.

1. Connectivity preserve modeler is suppose to preserve existing nodes if available and to replace elements with new destination elements. Therefore, it also should preserve the memory allocated on variables of nodes, making new modelpart to have same NodalSolutionStepVariablesList.
2. Current implementation checks whether the destination model part and origin model part has the same NodalSolutionStepVariablesList and throws a warning if not. So when it is used with a model part which is not a submodel part, it shows lots of warnings.
3. As per the model_parts, if destination model part is a submodel part of the origin model part, there is another check existing at this point, https://github.com/KratosMultiphysics/Kratos/blob/9bf51fac296325071ae356f511d094cb4345e866/kratos/sources/connectivity_preserve_modeler.cpp#L127-L138

Where it throws an error if submodelpart and origin model part NodalSolutionStepVariablesList differs, and if not it assigns the same NodalSolutionStepVariablesList of the origin to the destination modelpart. As we believe, this is the inteded behaviour since this is called "connectivity_preserving_modeler" which creates a new model part consist with same nodes.
4. This would make the following check block unnecessary, so we would suggest to remove this "CheckVariableLists" method completely.

https://github.com/KratosMultiphysics/Kratos/blob/9bf51fac296325071ae356f511d094cb4345e866/kratos/sources/connectivity_preserve_modeler.cpp#L92-L109

Is there any other intended behaviours expected by the "connectivity_preserve_modeler"? , Would removing the "CheckVariableLists" break the inteded behaviour?

@philbucher @RiccardoRossi @pooyan-dadvand @loumalouomega 

Thanks and best regards

**Comment by RiccardoRossi** [comment]
hi, i am not against using tje same variablelist, nevertheless this gives a memory managment problem for that.

also what woukd u do if you want to fill by the connectovoty preserve modeler an exosting modelpart? (with a variable list that is dofferent?)

that works ok i you have the same variables in tje list, but it is a prpvlem otherwise...so i think that the warning is ok ..

**Comment by RiccardoRossi** [comment]
maybe can u make a test showing what the problem is? it would be easier to follow...

**Comment by jcotela** [comment]
@RiccardoRossi I do not think that we need to change the behavior, it is just that we are effectively checking for something we don't really need. We start with two ModelParts, Source and Destination. We do not impose any restriction on them, so that they may be "root" ModelParts, SubModelParts on the same "tree" or completely unrelated SubModelParts. The logic of the checks as now is as follows:

1. Check if the VariablesList of the two modelparts match, throw a warning (the one in #4267)
2. Completely delete the Destination modelpart, which at this point has NO nodes.

Now we want to make the Destination modelpart contain the same nodes as the Source. Given this, the only sane option is ensuring that the two modelparts have the same VariablesList

3. Check if the Destination modelpart is a SubModelPart. If it is not, go to 5.
4. If it is a SubModelPart, check if it has the same VariablesList as the Source (this would work if they have a common parent ModelPart) -- But we already checked for that in point 1 (?!)
If the two do not match, throw an error, we cannot transfer the nodes.
5. If the destination is not a SubModelPart, replace the variables list (but then why did we bother checking in point 1?)

In short, the only point where it makes sense to check what is the Destination VariablesList is point 4, so the check in 1 is redundant.

Mentioning @ddiezrod and @pablobecker  since they were reporting a closely related issue in #4267

**Comment by RiccardoRossi** [comment]
ouch this is complex.

first of all, when you say "the same variablelist" you imply a pointer to the same object or two identical variable lists?

secondly, can it happen that you want to sunstitute an empty submodelpart within a non empty root? that would be a disaster ...

finally when you propose to substitute the variable lists, will u ensure that the same variable list is kept for a same root mp?

**Comment by RiccardoRossi** [comment]
in any case if this works as intended, tje warning would be silent...so we could keep it at least as a sanity check

**Comment by jcotela** [comment]
> ouch this is complex.
> 
> first of all, when you say "the same variablelist" you imply a pointer to the same object or two identical variable lists?
> 

For me "the same variablelist" should be a pointer to the same object. As we were discussing yesterday, I do not know if this is possible now, so I left it deliberately vague. I guess that two identical variable lists should work, but I could see ways of breaking that by updating a modelpart but not the other.

> secondly, can it happen that you want to sunstitute an empty submodelpart within a non empty root? that would be a disaster ...
> 

If you have the case of "empty submodelpart with non empty root", it should fail with an error (I understand this case would fail the check in 4 above, so it is covered by the current implementation -- never checked it though).

> finally when you propose to substitute the variable lists, will u ensure that the same variable list is kept for a same root mp?


If the two modelparts have the same root, they should have the same variable list by design, right?


> in any case if this works as intended, tje warning would be silent...so we could keep it at least as a sanity check

There are two warnings, one as an initial check (point 1) and another in the problem case of the non-empty root (point 4). The first one will complain even if the variables list can be safely replaced.

**Comment by RiccardoRossi** [comment]
ultimately i am ok with the solution of assigning the variablelists pointer.

i do believe that we need first to do memory managment for variablelists by intrusive_ptr bjt this is somewhat a detail.

my worry is that here details will matter a lot 😉
good thing is that we shall be able to test any of this changes pretty comprehensively...

**Comment by pooyan-dadvand** [comment]
hmmm, complex enough.... thanks to @sunethwarna and @jcotela for their clear explanation. 

lets put my logics:

```c++
if destination is root model_part:
    if is not parent of source
        // remove all entities and sub model parts if there are any and copy from source 
    else
        // error: please copy the elements/conditions and use replace elements/conditions
else
    if is sub of the source
        // error: changing the sub affect the source
    else
         // remove the entities and try to add the nodes and elements and send error if conflicts with parent model parts
```

Am I missing someting?

**Comment by RiccardoRossi** [comment]
just to tell that #4995 gives you the basis for experimenting if you like
